src/reclaim.rs

```diff
@@ -0,0 +1,186 @@
+pub(crate) use seize::{Collector, Guard, Linked};
+
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::sync::atomic::Ordering;
+use std::{fmt, ptr};
+
+pub(crate) struct Atomic<T>(seize::AtomicPtr<T>);
+
+impl<T> Atomic<T> {
+    pub(crate) fn null() -> Self {
+        Self(seize::AtomicPtr::default())
+    }
+
+    pub(crate) fn load<'g>(&self, _: Ordering, guard: &'g Guard<'_>) -> Shared<'g, T> {
+        guard.protect(&self.0).into()
+    }
+
+    pub(crate) fn store(&self, new: Shared<'_, T>, ordering: Ordering) {
+        self.0.store(new.ptr, ordering);
+    }
+
+    pub(crate) unsafe fn into_box(self) -> Box<Linked<T>> {
+        Box::from_raw(self.0.into_inner())
+    }
+
+    pub(crate) fn swap<'g>(
+        &self,
+        new: Shared<'_, T>,
+        ord: Ordering,
+        _: &'g Guard<'_>,
+    ) -> Shared<'g, T> {
+        self.0.swap(new.ptr, ord).into()
+    }
+
+    pub(crate) fn compare_exchange<'g>(
+        &self,
+        current: Shared<'_, T>,
+        new: Shared<'g, T>,
+        success: Ordering,
+        failure: Ordering,
+        _: &'g Guard<'_>,
+    ) -> Result<Shared<'g, T>, CompareExchangeError<'g, T>> {
+        match self
+            .0
+            .compare_exchange(current.ptr, new.ptr, success, failure)
+        {
+            Ok(ptr) => Ok(ptr.into()),
+            Err(current) => Err(CompareExchangeError {
+                current: current.into(),
+                new,
+            }),
+        }
+    }
+}
+
+impl<T> From<Shared<'_, T>> for Atomic<T> {
+    fn from(shared: Shared<'_, T>) -> Self {
+        Atomic(shared.ptr.into())
+    }
+}
+
+impl<T> Clone for Atomic<T> {
+    fn clone(&self) -> Self {
+        Atomic(self.0.load(Ordering::Relaxed).into())
+    }
+}
+
+impl<T> fmt::Debug for Shared<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:p}", self.ptr)
+    }
+}
+
+impl<T> fmt::Debug for Atomic<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:p}", self.0.load(Ordering::SeqCst))
+    }
+}
+
+pub(crate) struct CompareExchangeError<'g, T> {
+    pub(crate) current: Shared<'g, T>,
+    pub(crate) new: Shared<'g, T>,
+}
+
+pub(crate) struct Shared<'g, T> {
+    ptr: *mut Linked<T>,
+    _g: PhantomData<&'g ()>,
```

Comment:
If `seize::Linked` doesn't require the `'g`, should we just get rid of it from `Shared` as well?

Reply:
This ties the lifetime of the reference to that of the guard, which makes things a little safer. `Linked` is not equivalent to `Shared`, it's just a wrapper type with some extra metadata.